Feature/monitoring spring prometheus grafana blogpost (#635)
ordina-jworks/ordina-jworks.github.io: _posts/2020-10-31-monitoring-spring-prometheus-grafana.md

---
layout: post
authors: [kevin_govaerts]
title: 'Monitoring Spring boot with Prometheus and Grafana'
image: /img/2020-10-31-monitoring-spring-prometheus-grafana/thumbnail.jpg
tags: [Spring, Prometheus, Grafana, Docker]
category: Monitoring
comments: true
---

# Table of contents
{:.no_toc}

- TOC
{:toc}

----

# Introduction

> In a distributed landscape where we are working with microservices, serverless applications or just event-driven architecture as a whole, than observability, which comprises monitoring, logging, tracing and alerting, is an important architectural concern.  

Reviewer: ```suggestion
In a distributed landscape where we are working with microservices, serverless applications, or just event-driven architecture as a whole, observability, which comprises monitoring, logging, tracing, and alerting, is an important architectural concern.  
```

There are a few reasons why we want visibility in our highly distributed systems:  
- Issues will occur, even when our best employees have built it.
- Distributed systems generate distributed failures, which can be devastating when we are not prepared in advance.
- Reveal mistakes early, which is great for improvement and learning.
- It keeps us accountable.
- Reduce the mean time to resolution (MTTR).

In this blogpost I will explain the core concepts of Prometheus and Grafana.  

> In the last section I set up a demo project, so you can follow along and implement monitoring in     your own applications. 

Reviewer: ```suggestion
In the last section I set up a demo project, so you can follow along and implement monitoring in your own applications. 
```

# Prometheus

## What is Prometheus? 

> Prometheus is an open source monitoring system, originally developed by SoundCloud, it  has a large community behind it and can monitor nearly anything.  

Reviewer: ```suggestion
Prometheus, originally developed by SoundCloud is an open source and community-driven project that graduated from the Cloud Native Computing Foundation.
It can aggregate data from almost everything:
```

- Microservices
- Multiple languages
- Linux servers
- Windows servers

## Why do we need Prometheus?

> In our modern times of microservices, devops is becoming more and more complex and therefore needs automation.  

Reviewer: ```suggestion
In our modern times of microservices, DevOps is becoming more and more complex and therefore needs automation.  
```

</div> 

Even though Prometheus has its own UI to show graphs and metrics, we will be using Grafana as an extra layer on top of this webserver, to query and visualize our database.  

### Prometheus targets

#### What does is monitor ? 
{:.no_toc}

Prometheus monitors nearly anything. It could be Linux/windows server, apache server, single applications, services, etc.  
It monitors **units** on those targets like:
- CPU usage
- Memory/ Disk usage
- Request count
- Request durations
- Exceptions count

The units that we monitor are called metrics, which get saved into the Prometheus time-series database.  
Prometheus's metrics are formatted like a human-readable text file.

<div style="text-align: center;">
  <img alt="Prometheus endpoint actuator" src="/img/2020-10-31-monitoring-spring-prometheus-grafana/prometheus-endpoint.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div> 

In this file we can see that there is a "HELP" comment which describes what the metric is, and we have a "TYPE" which can be one of four metric-types: 
- Counter: how many times X happened (exceptions)
- Gauge: what is the current value of X now ? (disk usage, cpu etc)

> - Histogram: how long or how big ?

Reviewer: ```suggestion
- Histogram: how long or how big?
```

#### Exporters
{:.no_toc}

There are a number of libraries and servers which help in exporting existing metrics from third-party systems as Prometheus metrics.
You can have a look at the [exporters and integration tools](https://prometheus.io/docs/instrumenting/exporters/){:target="_blank" rel="noopener noreferrer"} here.    

On a side note, these tools are also available as Docker images, so we can use them inside kubernetes clusters.  
We can run an exporter docker image for a MySQL database as a side container inside the MySQL pod, connect to it and start translating data, to expose it on the metrics endpoint.

#### Monitoring our own application
{:.no_toc}

If we want to add our own instrumentation to our code, to know how many server resources our own application is using, how many requests it is handling or how many exceptions occurred, then we need to use one of the [client libraries](https://prometheus.io/docs/instrumenting/clientlibs/){:target="_blank" rel="noopener noreferrer"}.
These libraries will enable us to declare all the metrics we deem important in our application, and expose them on the metrics endpoint.

## Configuring Prometheus

> To instruct Prometheus on what it needs to scrape we create a **prometheus.yml** configuration file.  

Reviewer: ```suggestion
To instruct Prometheus on what it needs to scrape, we create a **prometheus.yml** configuration file.  
```

</div> 

In this configuration file we declare a few things:  
1. global configs, like how often it will scrape its targets.
2. we can declare [rule files](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/){:target="_blank" rel="noopener noreferrer"}, so when we meet a certain condition, we get an alert.
3. which services it needs to monitor.

In this example you can see that Prometheus will monitor two things: 
- Our Spring boot application
- Its own health

Prometheus expects the data of our targets to be exposed on the `/metrics` endpoint, unless otherwise declared in the `metrics_path` field.  

### Alerts
{:.no_toc}

With Prometheus, we have the possibility to get notified when metrics have reached a certain point, which we can declare in the `.rules` files. 
Prometheus has a component which is called the "Alertmanager", and it can send notifications over various channels like emails, slack, PagerDuty, etc.  

### Querying our data
{:.no_toc}

Since Prometheus saves all our data in a time series database, which is located on disk in a custom timeseries format, we need to use PromQL query language, if we want to query this database.  

> We can do this via the Prometheus WebUI, or we can use some more powerful visualization tools like Grafana, to do this for us.

Reviewer: ```suggestion
We can do this via the Prometheus WebUI, or we can use some more powerful visualization tools like Grafana.
```

Afterwards we can run the Prometheus image by running the following command: 

```
docker run -d -p 9090:9090 -v <PATH_TO_prometheus.yml_FILE>:/etc/prometheus/prometheus.yml prom/prometheus 
```
We mount the prometheus.yml config file into the Prometheus image and expose port 9090, to the outside of docker.

When this is up and running we can access the Prometheus webUI on `localhost:9090`.  

<div style="text-align: center;">
  <img alt="Prometheus UI" src="/img/2020-10-31-monitoring-spring-prometheus-grafana/prometheusUI.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div> 

When we navigate to Status > Targets, we can check if our connections are up and are correctly configured.

<div style="text-align: center;">
  <img alt="Prometheus target tab" src="/img/2020-10-31-monitoring-spring-prometheus-grafana/prometheus-target.PNG" width="auto" height="auto" target="_blank" class="image fit">
</div> 

## Setup Grafana

To run Grafana we will use the same approach as with Prometheus. 

We download and run the image from Docker-hub.  
```
docker run -d -p 3000:3000 grafana/grafana
```

> Now we can access the Grafana UI from `localhost:3000`, where enter "admin" as login and password.  

Reviewer: ```suggestion
Now we can access the Grafana UI from `localhost:3000`, where you can enter "admin" as login and password.  
```

We would need to work backwards over every service, all the way back to the stopped container, to find out what is causing the problem.

A better way would be to have a tool which:
- Constantly monitors all services
- Alerts system admins when something crashes
- Identifies problems before they occur

Prometheus is exactly that tool, it can identify memory usage, CPU usage, available disk space, etc.  
We can predefine certain thresholds about which we want to get notified.  

In our example it could have been that the memory of our failing server would have reached 70% memory usage for more than 1 hour, and could've sent an alert to our admins before the crash happened.  

## How it works

### Prometheus server

The server does the actual monitoring work, and it consists of 3 main parts:  
- Storage, which is a time series database.
- Data retrieval worker, which is pulling the data from our target services.
- Webserver, which accepts [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/){:target="_blank" rel="noopener noreferrer"} queries to get data from our DB.

<div style="text-align: center;">

>   <img alt="prometheus server" src="/img/2020-10-31-monitoring-spring-prometheus-grafana/prom-server.jpg" width="auto" height="auto" target="_blank" class="image fit">

Reviewer 2: Image looks a bit stretched.
```suggestion
  <img alt="prometheus server" src="/img/2020-10-31-monitoring-spring-prometheus-grafana/prom-server.jpg" style="max-width: 100%; height:auto" target="_blank" class="image">
```

</div> 

Even though Prometheus has its own UI to show graphs and metrics, we will be using Grafana as an extra layer on top of this webserver, to query and visualize our database.  

### Prometheus targets

> #### What does is monitor ? 

Reviewer: ```suggestion
#### What does it monitor? 
```
